File: computing/interview/design-patterns/behavioral/interpreter/csharp/InterpreterDemo.cs
using System;
using System.Collections.Generic;

namespace Interview.DesignPatterns.Behavioral.Interpreter
{
    public sealed class Context
    {
        private readonly Dictionary<string, bool> _variables = new();

        public void Assign(string variable, bool value)
        {
            _variables[variable] = value;
        }

        public bool Lookup(string variable)
        {
            if (!_variables.TryGetValue(variable, out var value))
            {
                throw new InvalidOperationException("Variable not found");
            }

            return value;
        }
    }

    public interface IExpression
    {
        bool Interpret(Context context);
    }

    public sealed class VariableExpression : IExpression
    {
        private readonly string _name;

        public VariableExpression(string name)
        {
            _name = name;
        }

        public bool Interpret(Context context) => context.Lookup(_name);
    }

    public sealed class AndExpression : IExpression
    {
        private readonly IExpression _left;
        private readonly IExpression _right;

        public AndExpression(IExpression left, IExpression right)
        {
            _left = left;
            _right = right;
        }

        public bool Interpret(Context context) => _left.Interpret(context) && _right.Interpret(context);
    }

    public sealed class OrExpression : IExpression
    {
        private readonly IExpression _left;
        private readonly IExpression _right;

        public OrExpression(IExpression left, IExpression right)
        {
            _left = left;
            _right = right;
        }

        public bool Interpret(Context context) => _left.Interpret(context) || _right.Interpret(context);
    }

    public sealed class NotExpression : IExpression
    {
        private readonly IExpression _operand;

        public NotExpression(IExpression operand)
        {
            _operand = operand;
        }

        public bool Interpret(Context context) => !_operand.Interpret(context);
    }

    public static class Demo
    {
        public static void Main()
        {
            var context = new Context();
            context.Assign("x", true);
            context.Assign("y", false);

            IExpression expression = new OrExpression(
                new AndExpression(new VariableExpression("x"), new VariableExpression("y")),
                new NotExpression(new VariableExpression("y")));

            Console.WriteLine($"Result: {expression.Interpret(context)}");
        }
    }
}

File: computing/interview/design-patterns/behavioral/interpreter/java/InterpreterDemo.java
package interview.designpatterns.behavioral.interpreter;

import java.util.HashMap;
import java.util.Map;

final class Context {
    private final Map<String, Boolean> variables = new HashMap<>();

    void assign(String variable, boolean value) {
        variables.put(variable, value);
    }

    boolean lookup(String variable) {
        Boolean value = variables.get(variable);
        if (value == null) {
            throw new IllegalArgumentException("Variable not found");
        }
        return value;
    }
}

interface Expression {
    boolean interpret(Context context);
}

final class VariableExpression implements Expression {
    private final String name;

    VariableExpression(String name) {
        this.name = name;
    }

    @Override
    public boolean interpret(Context context) {
        return context.lookup(name);
    }
}

final class AndExpression implements Expression {
    private final Expression left;
    private final Expression right;

    AndExpression(Expression left, Expression right) {
        this.left = left;
        this.right = right;
    }

    @Override
    public boolean interpret(Context context) {
        return left.interpret(context) && right.interpret(context);
    }
}

final class OrExpression implements Expression {
    private final Expression left;
    private final Expression right;

    OrExpression(Expression left, Expression right) {
        this.left = left;
        this.right = right;
    }

    @Override
    public boolean interpret(Context context) {
        return left.interpret(context) || right.interpret(context);
    }
}

final class NotExpression implements Expression {
    private final Expression operand;

    NotExpression(Expression operand) {
        this.operand = operand;
    }

    @Override
    public boolean interpret(Context context) {
        return !operand.interpret(context);
    }
}

public final class InterpreterDemo {
    private InterpreterDemo() {}

    public static void main(String[] args) {
        Context context = new Context();
        context.assign("x", true);
        context.assign("y", false);

        Expression expression = new OrExpression(
            new AndExpression(new VariableExpression("x"), new VariableExpression("y")),
            new NotExpression(new VariableExpression("y")));

        System.out.printf("Result: %b%n", expression.interpret(context));
    }
}
